### DISCUSSION TITLE
What is the easiest way to add a font size slider without using egui.settings?

### DISCUSSION THREAD

**Discussion by NadyaNayme** [emilk/egui#1478] [Q&A] [answered]
I'm a beginner to Rust - as in maybe a week into things now. Every existing issue/discussion I've found points to long out-of-date examples that have since changed. I've tried following the breadcrumbs and ended up short. I don't want to add my own fonts and I can't figure out how to persist the `egui.settings` font size sliders because I can't figure out how to access or set that value. All I want is for people to be able to bump up the much-to-small-by-default font size for accessibility reasons.

I'm sure I've passed over the documentation to do exactly what I want to do several dozen times now but at this point my brain has shut off. I've currently resorted to scaling the entire UI with `ctx.set_pixels_per_point` but this is less-than-desirable because it also scales my 32x32 icons and those become blurry when scaled so I want to do things properly.

![image](https://user-images.githubusercontent.com/4567166/162875428-536de7dc-f7a4-455c-aceb-c86d3f52e600.png)

**Comment by coderedart** [comment]
There's three things to remember.

1. [Font Family](https://docs.rs/egui/latest/egui/enum.FontFamily.html) which is either proportional (for normal text), monospace (for numbers in dragvalue ) or some custom thing with a custom name.
2. [Font Id](https://docs.rs/egui/latest/egui/struct.FontId.html) which is just a Font Family (above) + Font size.
3. [TextStyle](https://docs.rs/egui/latest/egui/style/enum.TextStyle.html) which is just an easy enum for small, monospace, heading, normal or any custom text styles

to set custom size for all text styles:
[Style](https://docs.rs/egui/latest/egui/style/struct.Style.html#) just has a `Map<TextStyle, FontID>`. so, you can just get style with `ctx.style()`, clone the `style`, set the sizes you want by 
```rust
for (_text_style, font_id) in style.text_styles.iter_mut() {
    font_id.size = 24 // whatever size you want here
}
```
you can remember user preferences for all default text styles     
```
   Small,
    Body,
    Monospace,
    Button,
    Heading
```
, and just set the sizes for those text styles with `style.text_styles.get_mut(TextStyle::Small).unwrap().size = 24`.

to instead just set a single font for everything, style also has a `override_font_id` field. just set that field with `style.override_font_id = Some(FontID::proportional(24)); // whatever size you want` to use that size as default for all widgets.

once you are done setting font size either way, you just need to tell egui to use that style with `ctx.set_style()` and that should work.

I can't really test that code at the moment, but that is the rough sketch. 

you can also just use `ctx.style_ui` to just use the egui provided UI, then serialize or deserialize the style into json or something to store it. that's how I do it in my app to support "user themes" which they can share with each other. (given they also share their fonts with each other, but you don't use custom fonts anyway)

**Reply by NadyaNayme** [reply]
I really do appreciate this writeup and it clarified with me that my understanding was fuzzy at the edges but mostly right.

This is roughly as far as I got but I couldn't get the code to work because I had no idea what to actually _do_ with `style.text_styles` or how I was meant to get the `TextStyle` I wanted to modify. I was messing around with some of the example code you gave to try it to compile and got stuck with a compiler issues regarding Arc and Deref - turns out I was meant to `*ctx.style()` as shown by emilks code. 

With my reading of this help text though - I would never in a million years had tried to deref the Arc because it was telling me it wasn't implemented for Arc - though I'm nearly positive I'm misreading or misunderstanding the message.

![image](https://user-images.githubusercontent.com/4567166/162904455-fc323c84-4844-447c-9ab7-e9d2c517e4e3.png)


However emilk's example code from below worked - swapping out the hard coded values with my config values using the below block:

```
    let mut style = (*ctx.style()).clone();
    style.text_styles = [
        (Heading, FontId::new(30.0, Proportional)),
        (Body, FontId::new(18.0, Proportional)),
        (Monospace, FontId::new(14.0, Proportional)),
        (Button, FontId::new(14.0, Proportional)),
        (Small, FontId::new(10.0, Proportional)),
    ]
    .into();
    ctx.set_style(style);
```

Works like a charm! My users thank you emilk :) 👍

**Comment by emilk** [comment, marked as answer]
There is an example for this coming in a PR: https://github.com/emilk/egui/pull/1476/files
